### PR TITLE
Feature #2562 version lookup documentation and automated tests

### LINKED ISSUE #2562: Internal: Add logic to define and parse METplus inter-component version dependencies
## Describe the New Feature ##
This topic was discussed during the METplus All Hands meeting on April 24, 2024 and documented in [this section](https://docs.google.com/document/d/1Lk6QRNMq9fKYP7wG0kppuhCV2cQbC1aoGzKzkFkGSMI/edit#heading=h.b7qwmd56ruup) of the METplus Engineering meeting notes. Each [METplus coordinated release](https://metplus.readthedocs.io/en/latest/Users_Guide/release-notes.html#metplus-components-release-note-links) consists of a collection of METplus component releases. The integration testing performed in METplus (for all METplus components) and METviewer (for METplus-Analysis components) needs to know which component versions correspond with each other.

Currently, METplus uses numerical offsets to compute corresponding component version numbers. METplus version X.Y corresponds to:
- MET version X+6.Y
- METviewer version X.Y
- METplotpy version X-3.Y
- METcalcpy version X-3.Y
- METdataio version X-3.Y

While this logic does work, it does NOT work for METexpress which releases on a different cycle. And these relationships may break in the future depending on future funding levels. Note also, that this logic is not currently well-implemented for handling METviewer dependencies.

The METplus team decided that it would prefer to move from a numerical offset solution to a table lookup. Recommend adding a Python file to the METplus repo to define a table for all prior METplus coordinated releases along with utility functions to parse that table. Also consider how these version dependencies relate to the manage-externals and avoid duplication of logic as much as possible.

- [x] Define a table with 7 columns (one each for METplus, MET, METviewer, METplotpy, METcalcpy, METdataio, and METexpress) and 1 row for each prior coordinated release.
- [x] Define a utility function that...
  - Takes as input a current component and version number.
  - Takes as input a requested METplus component name.
  - Takes as input a requested version "type", such as release version number (e.g. vX.Y.Z), branch name (e.g. main_vX.Y or develop), or DockerHub tag (e.g. MET-11.1-latest).
  - Provides as output the corresponding version name.
- Many questions remain:
  - How to get the latest bugfix number for the vX.Y release? Perhaps use a vX.Y-latest GitHub tag? And define a GitHub action to auto-update the vX.Y-latest tags?
  - When should develop be returned? And what about beta and rc releases?
  - What about DockerHub image names?
- [x] Once this functionality exists, update METplus to make use of it.
- [x] Remember to update the METplus Release Guide to include updating this coordinated release table.
- [X] Define a METviewer issues to retrieve this table lookup from the METplus develop branch and make good use of it. -- dtcenter/METviewer#554 was created to address this

The goal throughout is avoiding hard-coding METplus component version dependencies and instead define that logic in a single spot.

### Acceptance Testing ###
*List input data types and sources.*
*Describe tests required for new functionality.*

### Time Estimate ###
3 days?

### Sub-Issues ###
Consider breaking the new feature down into sub-issues.
- [ ] *Add a checkbox for each sub-issue here.*

### Relevant Deadlines ###
*List relevant project deadlines here or state NONE.*

### Funding Source ###
2782544 (NOAA Base) 

### Assignee ###
- [x] Select **engineer(s)** or **no engineer** required
- [ ] Select **scientist(s)** or **no scientist** required

### Labels ###
- [x] Select **component(s)**
- [x] Select **priority**
- [x] Select **requestor(s)**

### Projects and Milestone ###
- [x] Select **Repository** and/or **Organization** level **Project(s)** or add **alert: NEED CYCLE ASSIGNMENT** label
- [x] Select **Milestone** as the next official version or **Future Versions**

## Define Related Issue(s) ##
Consider the impact to the other METplus components.
- [ ] [METplus](https://github.com/dtcenter/METplus/issues/new/choose), [MET](https://github.com/dtcenter/MET/issues/new/choose), [METdataio](https://github.com/dtcenter/METdataio/issues/new/choose), [METviewer](https://github.com/dtcenter/METviewer/issues/new/choose), [METexpress](https://github.com/dtcenter/METexpress/issues/new/choose), [METcalcpy](https://github.com/dtcenter/METcalcpy/issues/new/choose), [METplotpy](https://github.com/dtcenter/METplotpy/issues/new/choose)
- [ ] Will need a METviewer issue at a minimum.

## New Feature Checklist ##
See the [METplus Workflow](https://metplus.readthedocs.io/en/latest/Contributors_Guide/github_workflow.html) for details.
- [ ] Complete the issue definition above, including the **Time Estimate** and **Funding source**.
- [ ] Fork this repository or create a branch of **develop**.
Branch name: `feature_<Issue Number>_<Description>`
- [ ] Complete the development and test your changes.
- [ ] Add/update log messages for easier debugging.
- [ ] Add/update unit tests.
- [ ] Add/update documentation.
- [ ] Add any new Python packages to the [METplus Components Python Requirements](https://metplus.readthedocs.io/en/develop/Users_Guide/overview.html#metplus-components-python-requirements) table.
- [ ] Push local changes to GitHub.
- [ ] Submit a pull request to merge into **develop**.
Pull request: `feature <Issue Number> <Description>`
- [ ] Define the pull request metadata, as permissions allow.
Select: **Reviewer(s)** and **Development** issues
Select: **Repository** level development cycle **Project** for the next official release
Select: **Milestone** as the next official version
- [ ] Iterate until the reviewer(s) accept and merge your changes.
- [ ] Delete your fork or branch.
- [ ] Close this issue.

### PR DESCRIPTION
**Pull request by georgemccabe**
## Change Summary
* Update Release Guides to include information on how to update the version lookup table
* Update Contributor's Guide to include information on how to use version lookup script
* Added stubs for Contributor's Guide descriptions of METplus utility scripts
* Updated automated use case test scripts to use version lookup script instead of relying on manage_externals
* Clean up some automated test scripts to satisfy SonarQube complaints

## Pull Request Testing ##

- [X] Describe testing already performed for these changes:</br>

* Ran a few use case groups that use METplotpy and METdataio to ensure that changes to automated test scripts still work
* Reviewed documentation edits for clarify and formatting

- [X] Recommend testing for the reviewer(s) to perform, including the location of input datasets, and any additional instructions:</br>

* Review Contributor's Guide updates: https://metplus.readthedocs.io/en/feature_2562_ext_version/Contributors_Guide/utilities.html
* Review Bugfix Release Guide updates: https://metplus.readthedocs.io/en/feature_2562_ext_version/Release_Guide/met_bugfix.html#update-version-lookup-table
* Review METplus Official Release Guide updates: https://metplus.readthedocs.io/en/feature_2562_ext_version/Release_Guide/metplus_official.html#add-next-version-to-lookup-table

- [X] Do these changes include sufficient documentation updates, ensuring that no errors or warnings exist in the build of the documentation? **[Yes]**

- [X] Do these changes include sufficient testing updates? **[Yes]**

- [X] Will this PR result in changes to the test suite? **[No]**</br>
If **yes**, describe the new output and/or changes to the existing output:</br>

- [X] Do these changes introduce new SonarQube findings? **[No]**</br>
If **yes**, please describe:

- [X] Please complete this pull request review by **11/8/2024**.</br>

## Pull Request Checklist ##
See the [METplus Workflow](https://metplus.readthedocs.io/en/latest/Contributors_Guide/github_workflow.html) for details.
- [X] Add any new Python packages to the [METplus Components Python Requirements](https://metplus.readthedocs.io/en/develop/Users_Guide/appendixA.html#metplus-components-python-packages) table.
- [X] For any new datasets, an entry to the [METplus Verification Datasets Guide](https://metplus.readthedocs.io/en/latest/Verification_Datasets/index.html).
- [X] Review the source issue metadata (required labels, projects, and milestone).
- [X] Complete the PR definition above.
- [X] Ensure the PR title matches the feature or bugfix branch name.
- [x] Define the PR metadata, as permissions allow.
Select: **Reviewer(s)** and **Development** issue
Select: **Milestone** as the version that will include these changes
Select: **Coordinated METplus-X.Y Support** project for bugfix releases or **METplus-Wrappers-X.Y.Z Development** project for official releases
- [x] After submitting the PR, select the :gear: icon in the **Development** section of the right hand sidebar. Search for the issue that this PR will close and select it, if it is not already selected.
- [ ] After the PR is approved, merge your changes. If permissions do not allow this, request that the reviewer do the merge.
- [ ] Close the linked issue and delete your feature or bugfix branch from GitHub.

### CODE REVIEW

File: docs/Contributors_Guide/utilities.rst [hunk 1]
```diff
@@ -0,0 +1,222 @@
...
+
+Get MET vX.Y.Z development version from beta version
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+::
+
+    $ component_versions.py -v 6.0.0-beta3 -o MET --no-get_dev_version
+    develop
+
+If the *--no-get_dev_version* argument is provided,
+an input version that includes -betaN or -rcN will return *develop*
```
Comment (jprestop) [inline docs/Contributors_Guide/utilities.rst]:
```suggestion
an input version that includes -betaN or -rcN will return *develop*.
```

**Comment by jprestop** [comment]
@georgemccabe I'm trying to wrap my head around this. I see in the bugfix documentation it says to check out develop, and I understand why that would be.  But, do we also want to update the main version of metplus/component_versions.py during a bugfix releease?

**Comment by jprestop** [comment]
@georgemccabe Also, I'll note that the add_next_version_to_lookup.rst file text assumes that the METplus components will always cut a X.0.0 release and then cut a X.1.0 release and then cut a X+1.0.0 release, which may or may not be the case, but it seems simplest to leave as-is, since that it is likely, an account for exceptions as we go along.  Is that what you and @JohnHalleyGotway were thinking too?

**Comment by georgemccabe** [comment]
@jprestop, I believe we discussed that any logic used to obtain version info would download the script from the develop branch to ensure it is maintained in one place. This idea sounds good in theory, but the details depend on how the script is eventually used and may need to change. If we start using this script and determine that we do need to have an updated list in the main_vX.Y branch, we would need to update these instructions to update the numbers in both main_vX.Y and develop.

> @georgemccabe I'm trying to wrap my head around this. I see in the bugfix documentation it says to check out develop, and I understand why that would be. But, do we also want to update the main version of metplus/component_versions.py during a bugfix releease?

**Comment by georgemccabe** [comment]
So far that has been the case and aligns with our current development cycle of 2 releases per year: X.0 and X.1. If this process changes, we will likely have to review more than just this instruction for accuracy.

> @georgemccabe Also, I'll note that the add_next_version_to_lookup.rst file text assumes that the METplus components will always cut a X.0.0 release and then cut a X.1.0 release and then cut a X+1.0.0 release, which may or may not be the case, but it seems simplest to leave as-is, since that it is likely, an account for exceptions as we go along. Is that what you and @JohnHalleyGotway were thinking too?

**Review by jprestop — APPROVED** [review]
Thanks for all of your work on this task and for the discussion. I have reviewed the documentation and ensured that all tests passed. I approve this request.

**Comment by jprestop** [comment]
@georgemccabe I hope you don't mind but I'm going to squash and merge this now, so I can pull these changes into the branch I am working on now.  I did not, however, delete your branch, in case you still need it for some reason.  Feel free to delete when you're ready.

**Pull request merged**